### PR TITLE
[flutter_tools] allow device classes to provide platform-specific interface for devFS Sync

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Part of the investigation of go/flutter-improving-devfs-reliability revealed an easy optimization for hot reload/restart on desktop devices. This should improve performance of hot restart by several hundred milliseconds and reduce memory usage of both the tool and application.

### CODE REVIEW

File: packages/flutter_tools/lib/src/desktop_device.dart [hunk 1]
```diff
@@ -214,3 +223,34 @@ class DesktopLogReader extends DeviceLogReader {
...
+        final File destination = _fileSystem.file(baseUri.resolveUri(uri));
+        if (!destination.parent.existsSync()) {
+          destination.parent.createSync(recursive: true);
+        }
+        if (devFSContent is DevFSFileContent) {
+          final File content = devFSContent.file as File;
+          content.copySync(destination.path);
+          continue;
+        }
+        destination.writeAsBytesSync(await devFSContent.contentsAsBytes());
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/desktop_device.dart:250]: Note to self: fall through is for `DevFSStringContent`/`DevFSByteContent`

**Review by jmagman — APPROVED** [review]
I wish we had more desktop integration tests...

**Comment by jonahwilliams** [comment]
We do have macOS hot reload/restart tests thankfully. I double check what the status is for linux/windows

File: packages/flutter_tools/lib/src/desktop_device.dart [hunk 2]
```diff
@@ -214,3 +223,34 @@ class DesktopLogReader extends DeviceLogReader {
     // Nothing to dispose.
   }
 }
+
+/// An implementation of a devFS writer which copies physical files for devices
+/// running on the same host.
+class DesktopDevFSWriter implements DevFSWriter {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/desktop_device.dart:229]: We also discussed using this same code for the iOS simulator (can be in a future PR).

File: packages/flutter_tools/lib/src/macos/macos_device.dart [hunk 3]
```diff
@@ -89,16 +92,19 @@ class MacOSDevices extends PollingDeviceDiscovery {
     @required MacOSWorkflow macOSWorkflow,
     @required ProcessManager processManager,
     @required Logger logger,
+    @required FileSystem fileSystem,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/macos/macos_device.dart:95]: g3 doesn't like this new `@required`, it's passing one into the context.

File: packages/flutter_tools/lib/src/linux/linux_device.dart [hunk 4]
```diff
@@ -67,13 +70,15 @@ class LinuxDevices extends PollingDeviceDiscovery {
   LinuxDevices({
     @required Platform platform,
     @required FeatureFlags featureFlags,
+    @required FileSystem fileSystem,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/linux/linux_device.dart:73]: Same, g3 doesn't like this new `@required`.

**Comment by jonahwilliams** [comment]
We could actually figure out if this is usable on Android by checking if the created devFS directory exists on the host file system

**Comment by jonahwilliams** [comment]
Updated to make this landable in g3, no longer enables the change immediately

**Pull request merged**